File: LeetCode/cpp/208.cpp
class Trie {
    struct TrieNode {
        vector<TrieNode *> next;
        bool is_end;

        TrieNode() : next(vector<TrieNode *>(26, nullptr)), is_end(false) {}
    };

    TrieNode *root;
public:
    Trie() {
        root = new TrieNode();
    }

    void insert(string word) {
        auto node = root;
        for (auto c:word) {
            if (!node->next[c - 'a'])
                node->next[c - 'a'] = new TrieNode();
            node = node->next[c - 'a'];
        }
        node->is_end = true;
    }

    bool search(string word, bool search = true) {
        auto node = root;
        for (auto p:word) {
            if (node->next[p - 'a'])
                node = node->next[p - 'a'];
            else
                return false;
        }
        return search ? node->is_end : true;
    }

    bool startsWith(string prefix) {
        return search(prefix, false);
    }
};

File: LeetCode/python3/208.py
class Trie:
    class TrieNode:
        def __init__(self):
            self.is_end = False
            self.next = [None for _ in range(26)]

    def __init__(self):
        self.root = self.TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for w in word:
            if not node.next[ord(w) - ord('a')]:
                node.next[ord(w) - ord('a')] = self.TrieNode()
            node = node.next[ord(w) - ord('a')]
        node.is_end = True

    def search(self, word: str, search: bool = True) -> bool:
        node = self.root
        for w in word:
            if node.next[ord(w) - ord('a')]:
                node = node.next[ord(w) - ord('a')]
            else:
                return False
        return node.is_end if search else True

    def startsWith(self, prefix: str) -> bool:
        return self.search(prefix, False)
